pkg/tecdsa/dkg/member.go

```diff
@@ -256,9 +259,382 @@ func (fm *finalizingMember) MarkInactiveMembers(
+	// TODO: Consider moving the whole implementation to the resultSubmitter
+	//       interface
+
+	// Chain rejects the result if it has less than 25% safety margin.
+	// If there are not enough signatures to preserve the margin, it does not
+	// make sense to submit the result.
+	signatureThreshold := sm.config.HonestThreshold + (sm.config.GroupSize-sm.config.HonestThreshold)/2
+	if len(signatures) < signatureThreshold {
+		return fmt.Errorf(
+			"could not submit result with [%v] signatures for signature threshold [%v]",
+			len(signatures),
+			signatureThreshold,
+		)
 	}
+
+	onSubmittedResultChan := make(chan uint64)
+
+	subscription := resultSubmitter.OnDKGResultSubmitted(
+		func(event *ResultSubmissionEvent) {
+			onSubmittedResultChan <- event.BlockNumber
+		},
+	)
+
+	returnWithError := func(err error) error {
+		subscription.Unsubscribe()
+		close(onSubmittedResultChan)
+		return err
+	}
+
+	alreadySubmitted, err := resultSubmitter.IsGroupRegistered(
+		result.GroupPublicKeyBytes,
+	)
+
+	if err != nil {
+		return returnWithError(
+			fmt.Errorf(
+				"could not check if the result is already submitted: [%v]",
+				err,
+			),
+		)
+	}
+
+	// Someone who was ahead of us in the queue submitted the result. Giving up.
+	if alreadySubmitted {
+		return returnWithError(nil)
+	}
+
+	// Wait until the current member is eligible to submit the result.
+	eligibleToSubmitWaiter, err := sm.waitForSubmissionEligibility(
+		blockCounter,
+		startBlockHeight,
+		sm.config.ResultPublicationBlockStep,
+	)
+	if err != nil {
+		return returnWithError(
+			fmt.Errorf("wait for eligibility failure: [%v]", err),
+		)
+	}
+
+	for {
+		select {
+		case blockNumber := <-eligibleToSubmitWaiter:
+			// Member becomes eligible to submit the result.
+			subscription.Unsubscribe()
+			close(onSubmittedResultChan)
+
+			sm.logger.Infof(
+				"[member:%v] submitting DKG result with public key [0x%x] and "+
+					"[%v] supporting member signatures at block [%v]",
+				sm.index,
+				result.GroupPublicKeyBytes,
+				len(signatures),
+				blockNumber,
+			)
+
+			return resultSubmitter.SubmitDKGResult(
+				sm.index,
+				result,
+				signatures,
+			)
+		case blockNumber := <-onSubmittedResultChan:
+			sm.logger.Infof(
+				"[member:%v] leaving; DKG result submitted by other member at block [%v]",
+				sm.index,
+				blockNumber,
+			)
+			// A result has been submitted by other member. Leave without
+			// publishing the result.
+			return returnWithError(nil)
+		}
+	}
+}
+
+// waitForSubmissionEligibility waits until the current member is eligible to
+// submit a result to the blockchain. First member is eligible to submit straight
+// away, each following member is eligible after pre-defined block step.
+func (sm *submittingMember) waitForSubmissionEligibility(
+	blockCounter chain.BlockCounter,
+	startBlockHeight uint64,
+	blockStep uint64,
+) (<-chan uint64, error) {
+	// T_init + (member_index - 1) * T_step
+	blockWaitTime := (uint64(sm.index) - 1) * blockStep
```

Comment:
Something to discuss with @pdyraga: Do we want to determine the first submitter in the same way as in beacon DKG?

Reply:
I added a TODO above the `setupEligibilityQueue` function to revisit it later and determine how we should choose which member submits DKG result first.